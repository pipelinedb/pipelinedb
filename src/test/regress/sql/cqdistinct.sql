SET debug_sync_stream_insert = 'on';

-- COUNT(DISTINCT ...)
CREATE CONTINUOUS VIEW test_distinct_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream;
CREATE CONTINUOUS VIEW test_distinct_sw_count AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');
CREATE CONTINUOUS VIEW test_distinct_sw_count_small AS SELECT COUNT(DISTINCT x::integer) FROM test_distinct_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '1 second');;

ACTIVATE test_distinct_count, test_distinct_sw_count, test_distinct_sw_count_small;

INSERT INTO test_distinct_count_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_count_stream (z, y, x) VALUES (null, null, null);

DEACTIVATE test_distinct_count, test_distinct_sw_count, test_distinct_sw_count_small;

SELECT * FROM test_distinct_count;
\d+ test_distinct_count_mrel0;

SELECT * FROM test_distinct_sw_count;
\d+ test_distinct_sw_count_mrel0;

SELECT pg_sleep(1);
SELECT * FROM test_distinct_sw_count_small;

DROP CONTINUOUS VIEW test_distinct_count;
DROP CONTINUOUS VIEW test_distinct_sw_count;
DROP CONTINUOUS VIEW test_distinct_sw_count_small;

-- DISTINCT / DISTINCT ON
CREATE CONTINUOUS VIEW test_distinct AS SELECT DISTINCT x::int, y::int - z::int FROM test_distinct_stream;
CREATE CONTINUOUS VIEW test_distinct_on AS SELECT DISTINCT ON (x::int, y::int - z::int) x::int, y::int, z::int FROM test_distinct_stream;
CREATE CONTINUOUS VIEW test_distinct_sw AS SELECT DISTINCT x::int, y::int - z::int FROM test_distinct_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');
CREATE CONTINUOUS VIEW test_distinct_on_sw AS SELECT DISTINCT ON (x::int, y::int - z::int) x::int, y::int, z::int FROM test_distinct_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');

ACTIVATE test_distinct, test_distinct_on, test_distinct_sw, test_distinct_on_sw;

INSERT INTO test_distinct_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_distinct_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_distinct_stream (z, y, x) VALUES (null, null, null);

DEACTIVATE test_distinct, test_distinct_on, test_distinct_sw, test_distinct_on_sw;

SELECT * FROM test_distinct ORDER BY x;
SELECT * FROM test_distinct_on ORDER BY x;
SELECT * FROM test_distinct_sw ORDER BY x;
SELECT * FROM test_distinct_on_sw ORDER BY x;

DROP CONTINUOUS VIEW test_distinct;
DROP CONTINUOUS VIEW test_distinct_on;
DROP CONTINUOUS VIEW test_distinct_sw;
DROP CONTINUOUS VIEW test_distinct_on_sw;
