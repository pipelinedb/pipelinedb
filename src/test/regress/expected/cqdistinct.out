SET debug_sync_stream_insert = 'on';
CREATE CONTINUOUS VIEW test_hll_count AS SELECT COUNT(DISTINCT x::integer) FROM test_hll_count_stream;
CREATE CONTINUOUS VIEW test_hll_sw_count AS SELECT COUNT(DISTINCT x::integer) FROM test_hll_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '10 seconds');
CREATE CONTINUOUS VIEW test_hll_sw_count_small AS SELECT COUNT(DISTINCT x::integer) FROM test_hll_count_stream WHERE (arrival_timestamp > clock_timestamp() - interval '1 second');;
ACTIVATE test_hll_count, test_hll_sw_count, test_hll_sw_count_small;
INSERT INTO test_hll_count_stream (x, y, z) VALUES (1, 1, '1'), (1, 1, '1'), (4, 4, '4'), (7, 7, '7'), (0, 0, '0'), (2, 2, '2'), (4, 4, '4'), (8, 8, '8'), (8, 8, '8'), (7, 7, '7'), (9, 9, '9'), (6, 6, '6'), (2, 2, '2'), (6, 6, '6'), (5, 5, '5'), (3, 3, '3'), (9, 9, '9'), (0, 0, '0'), (5, 5, '5'), (3, 3, '3');
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('6', 6, 6);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('7', 7, 7);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('4', 4, 4);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('1', 1, 1);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('5', 5, 5);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('8', 8, 8);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('3', 3, 3);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('9', 9, 9);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('2', 2, 2);
INSERT INTO test_hll_count_stream (z, y, x) VALUES ('0', 0, 0);
INSERT INTO test_hll_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_hll_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_hll_count_stream (x, y, z) VALUES (null, null, null);
INSERT INTO test_hll_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_hll_count_stream (z, y, x) VALUES (null, null, null);
INSERT INTO test_hll_count_stream (z, y, x) VALUES (null, null, null);
DEACTIVATE test_hll_count, test_hll_sw_count, test_hll_sw_count_small;
SELECT * FROM test_hll_count;
 count 
-------
    10
(1 row)

\d+ test_hll_count_mrel0;
                 Table "public.test_hll_count_mrel0"
 Column |  Type  | Modifiers | Storage  | Stats target | Description 
--------+--------+-----------+----------+--------------+-------------
 count  | bigint |           | plain    |              | 
 _0     | bytea  |           | extended |              | 

SELECT * FROM test_hll_sw_count;
 count 
-------
    10
(1 row)

\d+ test_hll_sw_count_mrel0;
                        Table "public.test_hll_sw_count_mrel0"
 Column |           Type           | Modifiers | Storage  | Stats target | Description 
--------+--------------------------+-----------+----------+--------------+-------------
 count  | bigint                   |           | plain    |              | 
 _1     | bytea                    |           | extended |              | 
 _0     | timestamp with time zone |           | plain    |              | 
Indexes:
    "test_hll_sw_count_mrel0__0_idx" UNIQUE, btree (_0)

SELECT pg_sleep(1);
 pg_sleep 
----------
 
(1 row)

SELECT * FROM test_hll_sw_count_small;
 count 
-------
     0
(1 row)

DROP CONTINUOUS VIEW test_hll_count;
DROP CONTINUOUS VIEW test_hll_sw_count;
DROP CONTINUOUS VIEW test_hll_sw_count_small;
